File: source/weha/wikianalysis/categorizer/EditorialComments.java
package mo.umac.wikianalysis.categorizer;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import mo.umac.wikianalysis.diff.token.BasicEdit;
import mo.umac.wikianalysis.diff.token.Insertion;
import mo.umac.wikianalysis.lexer.MediawikiScannerConstants;
import mo.umac.wikianalysis.lexer.WikiToken;

public class EditorialComments extends AbstractEditAction {

	private Set<String> editorialTemplates;
	
	{
		weight = 0.5;
	}
	
	public EditorialComments() {
		editorialTemplates = new HashSet<String>();
		editorialTemplates.add("{{fact");
		editorialTemplates.add("{{unreferenced");
		editorialTemplates.add("{{stub");
		this.be = null;
	}
	
	public EditorialComments(BasicEdit b) {
		this.be = new BasicEdit[1];
		this.be[0] = b;
	}
	
	public boolean isAction(BasicEdit edit) {
		
		if (!(edit instanceof Insertion))
			return false;
		else
		{
			Insertion insEdit = (Insertion) edit;
			WikiToken[] content = insEdit.getContent();
			
			for (int i = 0; i < content.length; i++)
			{
				if (content[i].kind == MediawikiScannerConstants.TEMPLATE_BEGIN &&
					editorialTemplates.contains(content[i].image.toLowerCase()))
					return true;
			}
		}
		
		return false;
	}

	@Override
	public List<AbstractEditAction> classify(List<BasicEdit> editList) {
		ArrayList<AbstractEditAction> ret = new ArrayList<AbstractEditAction>();
		ArrayList<BasicEdit> newEditList = new ArrayList<BasicEdit>();
		
		for (int i = 0; i < editList.size(); i++)
		{
			BasicEdit b = editList.get(i);
			if (this.isAction(b))
			{
				if (b instanceof Insertion)
				{
					Insertion ie = (Insertion) b;
					WikiToken[] content = ie.getContent();
					
					for (int j = 0; j < content.length; j++)
					{
						if (content[j].kind == MediawikiScannerConstants.TEMPLATE_END)
						{
							WikiToken[] actionContent = Arrays.copyOfRange(content, 0, j+1);
							WikiToken[] remainder = Arrays.copyOfRange(content, j+1, content.length);
							if (remainder.length > 0)
							{
								ie.setContent(actionContent);
								editList.add(new Insertion(ie.getPos()+j+1, remainder));
							}
							break;
						}
					}
					ret.add(new EditorialComments(ie));
				}
			}
			else {
				newEditList.add(b);
			}
		}
		
		editList.clear();
		editList.addAll(newEditList);
		
		return ret;
	}

	@Override
	public int lengthCount() {
		return 1;
	}

}

File: weha/source/weha/categorizer/EditorialComments.php
<?php
require_once('AbstractEditAction.php');

class EditorialComments extends AbstractEditAction {

	private $editorialTemplates;
	
	public function __construct($b = null) {
		if ($b === null)
		{
			$this->editorialTemplates = array();
			$this->editorialTemplates[] = "{{fact";
			$this->editorialTemplates[] = "{{unreferenced";
			$this->editorialTemplates[] = "{{stub";
			$this->editorialTemplates[] = "{{vgood-small";
			$this->be = null;
		}
		else
		{		
			$this->be = array();
			$this->be[] = $b;
		}
	}
	
	public function isAction($edit) {
		
		if (!($edit instanceof Insertion))
			return false;
		else
		{
			$insEdit = $edit;
			$content = $insEdit->getContent();
			
			for ($i = 0; $i < count($content); $i++)
			{
				if ($content[$i]->kind == 
					WikiLexerConstants::TEMPLATE_BEGIN &&
					in_array(strtolower($content[$i]->image), $this->editorialTemplates) )
					return true;
			}
		}
		
		return false;
	}

	public function classify(&$editList) {
		$ret = array();
		$newEditList = array();
		
		for ($i = 0; $i < count($editList); $i++)
		{
			$b = $editList[$i];
			if ($this->isAction($b))
			{
				if ($b instanceof Insertion)
				{
					$ie = $b;
					$content = $ie->getContent();
					
					for ($j = 0; $j < count($content); $j++)
					{
						if ($content[$j]->kind == 
							WikiLexerConstants::TEMPLATE_END)
						{
							$actionContent = array_slice($content, 0, $j+1);
							$remainder = array_slice($content, $j+1, count($content)-$j-1);
							if (count($remainder) > 0)
							{
								$ie->setContent($actionContent);
								$editList[] = new Insertion($ie->getPos()+$j+1, $remainder);
							}
							break;
						}
					}
					$ret[] = new EditorialComments($ie);
				}
			}
			else {
				$newEditList[] = $b;
			}
		}
		
		$editList = $newEditList;
		
		return $ret;
	}

}
